web3/providers/websocket.py

```diff
@@ -18,57 +16,43 @@
 RESTRICTED_WEBSOCKET_KWARGS = {'uri', 'loop'}
 
 
-def _start_event_loop(loop):
-    asyncio.set_event_loop(loop)
-    loop.run_forever()
-    loop.close()
-
-
-def _get_threaded_loop():
-    new_loop = asyncio.new_event_loop()
-    thread_loop = Thread(target=_start_event_loop, args=(new_loop,), daemon=True)
-    thread_loop.start()
-    return new_loop
-
-
 def get_default_endpoint():
     return os.environ.get('WEB3_WS_PROVIDER_URI', 'ws://127.0.0.1:8546')
 
 
-class PersistentWebSocket:
+class WebSocketConnection:
 
-    def __init__(self, endpoint_uri, loop, websocket_kwargs):
-        self.ws = None
+    def __init__(self, endpoint_uri, websocket_kwargs):
+        self.ws = dict()
         self.endpoint_uri = endpoint_uri
-        self.loop = loop
         self.websocket_kwargs = websocket_kwargs
 
     async def __aenter__(self):
-        if self.ws is None:
-            self.ws = await websockets.connect(
-                uri=self.endpoint_uri, loop=self.loop, **self.websocket_kwargs
+        thread_id = threading.get_ident()
+        if self.ws.get(thread_id) is None:
+            loop = asyncio.get_event_loop()
+            self.ws[thread_id] = await websockets.connect(
```

Comment:
You might consider using a [`threading.local`](https://docs.python.org/3.6/library/threading.html#thread-local-data) for this as I think it is the idiomatic way of doing per-thread data, handling the `thread_id` stuff for you under the hood.